### PR TITLE
UCP/Stream: fix memory leak if data was not received

### PR DESCRIPTION
**Pull request by evgeny-leksikov**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4721/ for details (Mellanox internal link).

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -544,6 +544,20 @@ void ucp_ep_destroy_internal(ucp_ep_h ep, const char *message)
 
 static void ucp_ep_disconnected(ucp_ep_h ep)
 {
+    ucp_recv_desc_t  *rdesc;
+
+    while (!ucs_queue_is_empty(&ep->stream_data)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: let's extract it to a function - e.g ucp_ep_stream_data_purge() or similar

File: test/gtest/ucp/test_ucp_stream.cc [hunk 2]
```diff
@@ -64,3 +64,75 @@ UCS_TEST_P(test_ucp_stream, send_recv_data) {
...
+        return test_ucp_stream::get_ctx_params();
+    }
+
+    virtual void init() {
+        /* Skip entities creation */
+        test_base::init();
+
+        const ucp_test_param &test_param = GetParam();
+
+        if ("\\self" == test_param.transports.front()) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_stream.cc]: can use skip_loopback() (see example in test_upc_wireup.cc)

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: In next PRs i'll extract stream_data and other stream specific stuff to special EP extension which is allocated/initialized and freed in corresponding functions only when this CAP requested. The data will be purged there if you dont have objection to keep it as is for now.

Comment (yosefe) [reply src/ucp/core/ucp_ep.c]: ok

File: test/gtest/ucp/test_ucp_stream.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_stream.cc]: Forgot, this test creates different entities connection map - many2one. sender == receiver check does not work here.

Comment (yosefe) [reply test/gtest/ucp/test_ucp_stream.cc]: how about adding function ucp_test::is_self(), and use it from ucp_test::init() as well?

Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_stream.cc]: np

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4728/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
#1884 
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4730/ for details (Mellanox internal link).

File: test/gtest/ucp/ucp_test.cc [hunk 3]
```diff
@@ -55,14 +55,18 @@ void ucp_test::cleanup() {
...
-
     create_entity();
-    if ("\\self" != test_param.transports.front()) {
+    if (!is_self()) {
         create_entity();
     }
 }
 
+bool ucp_test::is_self() const {
+    const ucp_test_param &test_param = GetParam();
```
Comment (yosefe) [inline test/gtest/ucp/ucp_test.cc]: is this local variable is not really needed?

**Comment by evgeny-leksikov** [comment]
@yosefe thanks! i'll rebase and rename commit messages after review.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4737/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by yosefe** [comment]
@evgeny-leksikov can you please squash all commits (and use the correct commit message formatting in the resulting commit)? 10x!

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4739/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4743/ for details (Mellanox internal link).

**Pull request merged**